x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/process/MlMemoryTracker.java

```diff
@@ -411,29 +427,31 @@ public void refresh(PersistentTasksCustomMetadata persistentTasks, ActionListene
                 refreshComplete::onFailure
             );
 
-            List<PersistentTasksCustomMetadata.PersistentTask<?>> mlAnomalyDetectorJobTasks = persistentTasks.tasks()
-                .stream()
-                .filter(task -> MlTasks.JOB_TASK_NAME.equals(task.getTaskName()))
-                .collect(Collectors.toList());
-            iterateAnomalyDetectorJobTasks(mlAnomalyDetectorJobTasks.iterator(), refreshDataFrameAnalyticsJobs);
+            Set<String> mlAnomalyDetectorJobTasks = Stream.concat(
+                persistentTasks.tasks()
+                    .stream()
+                    .filter(task -> MlTasks.JOB_TASK_NAME.equals(task.getTaskName()))
+                    .map(task -> MlTasks.jobId(task.getId())),
+                persistentTasks.tasks()
+                    .stream()
+                    .filter(task -> MlTasks.JOB_SNAPSHOT_UPGRADE_TASK_NAME.equals(task.getTaskName()))
+                    .map(task -> MlTasks.jobAndSnapshotId(task.getId()).v1())
```

Comment:
Why can't you cast the task.getParams() to `SnapshotUpgradeTaskParams` and grab the job ID that way?

Reply:
Good idea.  It's a better way to do it for all the types of tasks.